blog/2024-04-16-custom-components.md

```diff
@@ -0,0 +1,198 @@
+
+```bash
+pip install reflex-chat
+```
+
+You can then import the chat component into your app and use it like any other Reflex component. All the styling and logic is encapsulated, you only need to pass in the actual LLM logic.
+
+```python exec
+import reflex as rx
+from openai import OpenAI
+from reflex_chat import chat
+
+try:
+    client = OpenAI()
+except:
+    client = None
+
+# Only define your logic, the chat component handles the rest.
+async def run_llm(chat_state):
+    # Start a new session to answer the question.
+    print(chat_state.messages)
+    session = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=chat_state.messages,
+        stream=True,
+    )
+
+    # Stream the results, yielding after every word.
+    for item in session:
+        if hasattr(item.choices[0].delta, "content"):
+            answer_text = item.choices[0].delta.content
+            # Ensure answer_text is not None before concatenation
+            chat_state.append_to_response(answer_text)
+            yield
+```
+
+```python demo
+rx.box(
+    chat(process=run_llm),
+    height="500px",
+    width="100%"
+)
+```
+
+```md alert info
+# Just set `process` prop of the chat component to connect to your LLM.
+
+\```python
+async def run_llm(chat_state):
+    # Start a new session to answer the question.
+    session = client.chat.completions.create(
+        model="gpt-3.5-turbo",
+        messages=chat_state.messages,
+        stream=True,
+    )
+
+    # Stream the results, yielding after every word.
+    for item in session:
+        if hasattr(item.choices[0].delta, "content"):
+            answer_text = item.choices[0].delta.content
+            chat_state.append_to_response(answer_text)
+            yield
+\```
+```
+
+Depending on how much control you want, you can either use a high level component directly or create your own component from scratch.
+
+### Component State
+
+A new feature we've added in 0.4.6 is the `rx.ComponentState` class. This class allows you to encapsulate state and UI for a component in a single class. In fact, this is what allows the `chat` component to work without having the user having to define a state - the component will dynamically create and handle its own state internally.
+
+For example we can create a counter component that increments and decrements a count.
+
+```python demo exec
+class Counter(rx.ComponentState):
+    count: int = 0
+
+    def increment(self):
+        self.count += 1
+
+    def decrement(self):
+        self.count -= 1
+
+    @classmethod
+    def get_component(cls, **props) -> rx.Component:
+        return rx.hstack(
+            rx.button("Decrement", on_click=cls.decrement),
+            rx.heading(cls.count),
+            rx.button("Increment", on_click=cls.increment),
+        )
+
+counter = Counter.create
+```
+
+Then we can create multiple instances of the counter, each with its own state.
+
```

Comment:
can add this to hack around the counters being tied together when they shouldn't be

    ```python exec
    # Hack because flexdown re-inits modules
    Counter._per_component_state_instance_count = 2
    ```

Then down below the next block

    ```python exec
    # Hack because flexdown re-inits modules
    Counter._per_component_state_instance_count = 5
    ```

Then they will actually operate independently